File: dotnet/src/SimpleOrm/MappingException.cs
using System.Text;

namespace SimpleOrm;

/// <summary>
/// Thrown when loading an <see cref="EntityMap"/> finds violations. Carries every
/// violation found for the type — loaders never stop at the first error.
/// </summary>
public sealed class MappingException : Exception
{
    public MappingException(Type entityType, IReadOnlyList<MappingError> errors)
        : base(BuildMessage(entityType, errors))
    {
        EntityType = entityType;
        Errors = errors;
    }

    public Type EntityType { get; }

    public IReadOnlyList<MappingError> Errors { get; }

    private static string BuildMessage(Type entityType, IReadOnlyList<MappingError> errors)
    {
        var builder = new StringBuilder()
            .Append("Mapping of '").Append(entityType.FullName).Append("' failed with ")
            .Append(errors.Count).Append(errors.Count == 1 ? " error:" : " errors:");
        foreach (var error in errors)
        {
            builder.AppendLine().Append("  ").Append(error);
        }

        return builder.ToString();
    }
}

File: php/src/Errors/MappingException.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Errors;

/**
 * Every violation found while loading one type (spec/metadata-model.md: never
 * first-error-only). The exception's own code is the first violation's, so
 * `$e->errorCode` still names a rule; `$errors` carries the complete list.
 */
final class MappingException extends SimpleOrmException
{
    /** @param non-empty-list<MappingError> $errors */
    public function __construct(
        string $entityType,
        public readonly array $errors,
    ) {
        parent::__construct(
            $errors[0]->code,
            $entityType,
            count($errors) === 1
                ? $errors[0]->message
                : count($errors) . " mapping errors:\n  " . implode("\n  ", array_map('strval', $errors)),
        );
    }
}
